File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/gzgets.c
char * ZEXPORT gzgets(gzFile file, char *buf, int len) {
    unsigned left, n;
    char *str;
    unsigned char *eol;
    gz_statep state;

    /* check parameters and get internal structure */
    if (file == NULL || buf == NULL || len < 1)
        return NULL;
    state = (gz_statep)file;

    /* check that we're reading and that there's no (serious) error */
    if (state->mode != GZ_READ ||
        (state->err != Z_OK && state->err != Z_BUF_ERROR))
        return NULL;

    /* process a skip request */
    if (state->seek) {
        state->seek = 0;
        if (gz_skip(state, state->skip) == -1)
            return NULL;
    }

    /* copy output bytes up to new line or len - 1, whichever comes first --
       append a terminating zero to the string (we don't check for a zero in
       the contents, let the user worry about that) */
    str = buf;
    left = (unsigned)len - 1;
    if (left) do {
        /* assure that something is in the output buffer */
        if (state->x.have == 0 && gz_fetch(state) == -1)
            return NULL;                /* error */
        if (state->x.have == 0) {       /* end of file */
            state->past = 1;            /* read past end */
            break;                      /* return what we have */
        }

        /* look for end-of-line in current output buffer */
        n = state->x.have > left ? left : state->x.have;
        eol = (unsigned char *)memchr(state->x.next, '\n', n);
        if (eol != NULL)
            n = (unsigned)(eol - state->x.next) + 1;

        /* copy through end-of-line, or remainder if not found */
        memcpy(buf, state->x.next, n);
        state->x.have -= n;
        state->x.next += n;
        state->x.pos += n;
        left -= n;
        buf += n;
    } while (left && eol == NULL);

    /* return terminated string, or if nothing, end of file */
    if (buf == str)
        return NULL;
    buf[0] = 0;
    return str;
}

File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/gzgets.rs
rust
// This is an example of how to translate C code to Rust using the rust-bindgen tool.

use std::ffi::CString;
use std::os::raw::c_char;

fn gzgets(file: *mut GZFile, buf: *mut c_char, len: u32) {
    let state = unsafe { (file as *mut GZState).as_ref() }.unwrap();

    if state.mode != GZMode::Read || state.err != ZErrno::Ok && state.err != ZErrno::BufError {
        return;
    }

    if state.seek {
        state.seek = 0;
        if gz_skip(state, state.skip) == -1 {
            return;
        }
    }

    let mut str = buf;
    let mut left = len as u32 - 1;
    while left > 0 {
        if state.x.have == 0 && gz_fetch(state) == -1 {
            return;
        }

        if state.x.have == 0 || state.past {
            break;
        }

        let mut n = state.x.have as u32 > left ? left : state.x.have as u32;
        let eol = unsafe { memchr(state.x.next, '\n', n) };
        if eol.is_some() {
            n += 1;
        }

        unsafe {
            std::ptr::copy_nonoverlapping(state.x.next, buf, n as usize);
            state.x.have -= n as u32;
            state.x.next += n as usize;
            state.x.pos += n as u64;
        }

        left -= n;
        str = unsafe { buf.offset(n as isize) };
    }
}